implementation/matmul-rst-hackweek/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+use anyhow::Error;
+use num::traits::Zero;
+use std::{
+    cmp::{PartialEq, PartialOrd},
+    fmt::{Debug, Display},
+    ops::{Add, AddAssign, Div, Mul, Neg, Sub},
+    vec::Vec,
+};
+use thiserror::Error;
+
+// TODO This determines whether we transpose the second matrix first before multiplying. This
+// should pay off for large matrices, but I haven't tested what the actual threshold should be.
+const TRANSPOSE_THRESHOLD: usize = 1e9 as usize;
+
+pub trait AsAnyhowError: From<AsStdError> + Send + Sync {}
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub struct AsStdError(#[from] anyhow::Error);
+impl AsAnyhowError for AsStdError {}
+
+/// Matrix interface, generic over the the type of the elements contained within the matrix.
+/// The element type must be an implementer of `Element`.
+pub trait Matrix<E>: Clone
+where
+    E: Element,
+{
+    type MatrixError: Debug + AsAnyhowError + From<AsStdError> + From<anyhow::Error>;
+
+    fn from_vec(num_rows: usize, num_cols: usize, data: &Vec<E>)
+        -> Result<Self, Self::MatrixError>;
+
+    fn transpose(&self) -> Self;
+
+    fn matmul(&self, other: &Self) -> Self;
+}
+
+/// Collection of traits required by the elements of a Matrix.
+pub trait Element:
+    Debug
+    + Clone
+    + PartialEq
+    + Display
+    + Add<Output = Self>
+    + AddAssign
+    + Sub<Output = Self>
+    + Mul<Output = Self>
+    + Div<Output = Self>
+    + Zero
+{
+}
+
+// Below are some implementations of `Element` and `RealElement` "for free". This should facilitate
+// unit testing with these types.
+impl Element for usize {}
+impl Element for u32 {}
+impl Element for u16 {}
+impl Element for i32 {}
+impl Element for f64 {}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct MatrixImpl<E>
+where
+    E: Element,
+{
+    pub num_cols: usize,
+    pub num_rows: usize,
+    data: Vec<E>,
+}
+
+impl<E> MatrixImpl<E>
+where
+    E: Element + PartialOrd + Neg<Output = E>,
+{
+    pub fn compare(&self, other: &Self, tolerance: E) -> bool {
+        if self.num_cols != other.num_cols || self.num_rows != other.num_rows {
+            return false;
+        }
+        let neg_tolerance = -tolerance.clone();
+        for i in 0..self.data.len() {
+            let diff = self.data[i].clone() - other.data[i].clone();
+            if (tolerance < diff) || (diff < neg_tolerance) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
+impl<E: Element> MatrixImpl<E> {
+    /// Multiply self by the transpose of a matrix.
+    ///
+    /// This is equivalent to `self.matmul(other.transpose())`, but faster.
+    fn matmul_transpose(&self, other: &Self) -> MatrixImpl<E> {
+        debug_assert!(self.num_cols != other.num_cols);
+        let dim1 = self.num_rows;
+        let dim_inner = self.num_cols;
+        let dim2 = other.num_rows;
+        // Create an unallocated data vector for the result.
```

Comment:
```suggestion
        // Create an uninitialised data vector for the result.
```
```suggestion
        // Create an pre-allocated data vector for the result.
```
This looks like a typo to me: the vector is intentionally being allocated in advance.